content/articles/understanding-shared-memory-programming-with-pthreads-and-openmp/index.md

```diff
@@ -0,0 +1,246 @@
+
+    pthread_exit(NULL);
+}
+```
+The output of the code will be:
+
+"`bash
+New thread developed (3179116288)... 
+
+Hi and welcome to Section Engineering education - got 11 !
+```
+
+To run the above code in the UNIX system, run the below command:
+
+"`bash
+gcc first.c -lpthread -o output1
+```  
+
+Gcc is the software used to run the c++ codes in Linux systems. The `-lpthred` in the command tells the command that the file being compiled is a pthread program. The `-o' specifies the output of the compiled program.
+
+After the code is compiled, an output file is created and saved in the terminal's same directory folder. For our case, the output file will be saved as `output1` as we specified in the command. To view the output, run the below code:
+
+```bash 
+./output1
+```
+
+The below screenshot shows all the processes involved.
+
+![Process illustration](/engineering-education/understanding-shared-memory-programming-with-pthreads-and-openmp/process.png)
+
+#### Uses of pthreads
+Pthreads are useful in the following ways:
+
+1. `Build an adaptive user interface` - Threads are useful in user interfaces. Input from the user is processed and displayed in a loop. Processing might take longer in some instances, causing the user to wait. Using a separate thread for long-running tasks may improve software responsiveness. An operation may be halted.
+
+2. `Build a web server` - A web server must be able to download large files rapidly. Opening a new thread for each request saves time. Multiple threads can run on many processors at once.
+
+3. `Building a graphical user interface` - In graphical apps, a windowing system notification asking an app to redo a window portion is expected. Its window will be blank if it is preoccupied. In this scenario, having one thread handle windowing system messages and requests is prudent (as well as user input). If an operation takes more than 0.2 seconds, it is sent to another thread.
+
+### OpenMP and its use
+OpenMP is an SMP programming package (symmetric multi-processors). Threads in OpenMP exchange memory and data. This includes C++ and Fortran. OpenMP functions are in omp.h. There are sequential and parallel elements to an OpenMP program. It generally entails initializing variables and the environment.
+
+While an OpenMP application may use several threads simultaneously, only one is used sequentially (in the parallel sections). Only the controller thread goes end to end. When software is dismantled, more threads arise. 
+
+This specifies a parallel chunk of code (omp pragma). Whenever a parallel segment is reached, slave threads are established. The parallel code may run on another thread. Threads finish and join the master. Continue until all parallel threads are finished. An omp get thread num() method may get each thread's ID. The controller thread has ID 0. However, OpenMP masks the low-level issues and permits high-level parallel code definition.

```

Comment:
There seem to be some incomplete statements here. Would you please check it out?